test/hotspot/jtreg/compiler/arraycopy/stress/AbstractStressArrayCopy.java

```diff
@@ -0,0 +1,152 @@
+            len + " elements, at pos " + pos);
+    }
+
+    public static void throwContentsError(int l, int r, int len, int pos) {
+        throwError("in contents", l, r, len, pos);
+    }
+
+    public static void throwHeadError(int l, int r, int len, int pos) {
+        throwError("in head", l, r, len, pos);
+    }
+
+    public static void throwTailError(int l, int r, int len, int pos) {
+        throwError("in tail", l, r, len, pos);
+    }
+
+    private static void throwError(String phase, int l, int r, int len, int pos) {
+        throw new RuntimeException("Error " + phase + ": " +
+            len + " elements, " +
+            "[" + l + ", " + (l+len) + ") -> " +
+            "[" + r + ", " + (r+len) + "), " +
+            "at pos " + pos);
+    }
+
+    protected abstract void testWith(int size, int l, int r, int len);
+
+    public void exhaustiveWith(int size) {
+        for (int l = 0; l < size; l++) {
+            for (int r = 0; r < size; r++) {
+                int maxLen = Math.min(size - l, size - r);
+                for (int len = 0; len <= maxLen; len++) {
+                    testWith(size, l, r, len);
+                }
+            }
+        }
+    }
+
+    public void fuzzWith(Random rand, int size) {
+        // Some basic checks first
+        testWith(size, 0, 1, 1);
+        testWith(size, 0, 1, size-1);
+
+        // Test disjoint:
+        for (int c = 0; c < FUZZ_COUNT; c++) {
+            int l = rand.nextInt(size / 2);
+            int len = rand.nextInt((size - l) / 2);
+            int r = (l + len) + rand.nextInt(size - 2*len - l);
+
+            if (l >= size)      throw new IllegalStateException("l is out of bounds");
+            if (l + len > size) throw new IllegalStateException("l+len is out of bounds");
+            if (r >= size)      throw new IllegalStateException("r is out of bounds");
+            if (r + len > size) throw new IllegalStateException("r+len is out of bounds: " + l + " " + r + " " + len + " " + size);
+            if (l + len > r)    throw new IllegalStateException("Not disjoint");
+
+            testWith(size, l, r, len);
+            testWith(size, r, l, len);
+        }
+
+        // Test conjoint:
+        for (int c = 0; c < FUZZ_COUNT; c++) {
+            int l = rand.nextInt(size);
+            int len = rand.nextInt(size - l);
+            int r = Math.min(l + (len > 0 ? rand.nextInt(len) : 0), size - len);
+
+            if (l >= size)      throw new IllegalStateException("l is out of bounds");
+            if (l + len > size) throw new IllegalStateException("l+len is out of bounds");
+            if (r >= size)      throw new IllegalStateException("r is out of bounds");
+            if (r + len > size) throw new IllegalStateException("r+len is out of bounds: " + l + " " + r + " " + len + " " + size);
+            if (l + len < r)    throw new IllegalStateException("Not conjoint");
+
+            testWith(size, l, r, len);
+            testWith(size, r, l, len);
```

Comment:
The checks and testWith invocations of disjoint and conjoint are almost same except of "Not disjoint" and "Not conjoint" assert, it could be consolidated.

Reply:
Yes, good suggestion. See new commit.